### PR TITLE
FixConflictExamAPI

### PR DESCRIPTION
**Pull request by LinhNguyen3**
Fix Conflict

### CODE REVIEW

File: WIP/Source/Veazy/src/main/java/vn/edu/fu/veazy/core/model/ExamModel.java [hunk 1]
```diff
@@ -21,25 +23,30 @@
...
     private Integer userId;
     @Column(name = "courseId", nullable = false)
     private Integer courseId;
+    @Column(name = "questionSkill", nullable = true)
+    private Integer questionSkill;
     @OneToMany(mappedBy = "question")
     @Column(name = "listQuestions", nullable = false)
-    private List<ExamAnswer> listQuestions;
-    @Column(name = "result", columnDefinition="FLOAT4 DEFAULT 0.0",nullable = false)
+    private List<ExamAnswer> listQuestions = new ArrayList<>();
```
Comment (capheden) [inline WIP/Source/Veazy/src/main/java/vn/edu/fu/veazy/core/model/ExamModel.java:43]: Cho nay ong sang ben QuestionModel copy ve t sua roi day'.

File: WIP/Source/Veazy/src/main/java/vn/edu/fu/veazy/core/model/ExamAnswer.java [hunk 2]
```diff
@@ -24,27 +26,31 @@
...
     @ManyToOne
-    @JoinColumn(name="exammodel_id")
-    private ExamModel question;
+    @JoinColumn(name = "exammodel_id")
+    private ExamModel exam;
     @Column(name = "questionId", nullable = false)
     private Integer questionId;
     @Column(name = "userAnswer", nullable = false)
-    private String userAnswer;
+    private List<String> userAnswer = new ArrayList<>();
```
Comment (capheden) [inline WIP/Source/Veazy/src/main/java/vn/edu/fu/veazy/core/model/ExamAnswer.java:39]: @ElementCollection
@Access(AccessType.PROPERTY)
Xem ben QuestionModel

File: WIP/Source/Veazy/src/main/java/vn/edu/fu/veazy/core/controller/ExamController.java [hunk 3]
```diff
@@ -84,7 +94,163 @@ String createExam(@ModelAttribute("create-exam-form") CreateExamForm form,
...
+
+    /**
+     *
+     * @param examId
+     * @param principal
+     * @return json string
+     */
+    @RequestMapping(value = Const.URLMAPPING_GET_EXAM, method = RequestMethod.GET)
+    public @ResponseBody
+    String redoExam(@PathVariable("exam_id") Integer examId,
```
Comment (capheden) [inline WIP/Source/Veazy/src/main/java/vn/edu/fu/veazy/core/controller/ExamController.java:228]: Sai URLMAPPING

**Comment by capheden** [comment]
Sua lai font chu Nhat nhe.
Voi them @PreAuthorize(...) vao dau moi controller method. Xem vd trong may cai controller t sua roi ay.

File: WIP/Source/Veazy/src/main/java/vn/edu/fu/veazy/core/controller/ExamController.java [hunk 4]
```diff
@@ -84,7 +96,166 @@ String createExam(@ModelAttribute("create-exam-form") CreateExamForm form,
...
+    }
+
+    /**
+     *
+     * @param examId
+     * @param principal
+     * @return json string
+     */
+    @PreAuthorize("!isAuthenticated()")
+    @RequestMapping(value = Const.URLMAPPING_REDO_EXAM, method = RequestMethod.GET)
```
Comment (capheden) [inline WIP/Source/Veazy/src/main/java/vn/edu/fu/veazy/core/controller/ExamController.java:226]: @PreAuthorize("isAuthenticated()") // phai login thi moi redo duoc chu
or
@PreAuthorize("hasRole(3)") // cu the hon thi phai la learner (not editor || admin) moi dc redo

File: WIP/Source/Veazy/src/main/java/vn/edu/fu/veazy/core/controller/ExamController.java [hunk 5]
```diff
@@ -84,7 +96,166 @@ String createExam(@ModelAttribute("create-exam-form") CreateExamForm form,
...
+    }
+
+    /**
+     *
+     * @param examId
+     * @param principal
+     * @return json string
+     */
+    @PreAuthorize("!isAuthenticated()")
+    @RequestMapping(value = Const.URLMAPPING_GET_EXAM, method = RequestMethod.GET)
```
Comment (capheden) [inline WIP/Source/Veazy/src/main/java/vn/edu/fu/veazy/core/controller/ExamController.java:184]: tuong tu ben duoi

File: WIP/Source/Veazy/src/main/java/vn/edu/fu/veazy/core/controller/ExamController.java [hunk 6]
```diff
@@ -84,7 +96,166 @@ String createExam(@ModelAttribute("create-exam-form") CreateExamForm form,
...
+    }
+
+    /**
+     *
+     * @param form form submitted
+     * @param principal
+     * @return json string
+     */
+    @PreAuthorize("!isAuthenticated()")
+    @RequestMapping(value = Const.URLMAPPING_SUBMIT_EXAM_ANSWER, method = RequestMethod.POST)
```
Comment (capheden) [inline WIP/Source/Veazy/src/main/java/vn/edu/fu/veazy/core/controller/ExamController.java:117]: Sua thanh @PreAuthorize("permitAll()") // chap nhan moi ket noi && gan gia tri principle (neu co)
Ben duoi check principle == null la anonymous, != null thi moi luu ket qua (examService.save())

File: WIP/Source/Veazy/src/main/java/vn/edu/fu/veazy/core/controller/ExamController.java [hunk 7]
```diff
@@ -51,13 +60,16 @@
...
+    @Autowired
+    private QuestionService questionService;
 
     /**
      *
      * @param form form submitted
      * @param principal
      * @return json string
      */
+    @PreAuthorize("!isAuthenticated()")
```
Comment (minhs2linh) [inline WIP/Source/Veazy/src/main/java/vn/edu/fu/veazy/core/controller/ExamController.java]: tuong tu ben duoi

**Pull request merged**